### PR TITLE
proxy port range allocation

### PR DESCRIPTION
**Pull request by jdef**
implements changes proposed in #7846

/cc @thockin

### CODE REVIEW

**Comment by jdef** [comment]
I signed it!

I work for Mesosphere

File: cluster/addons/dns/kube2sky/RELEASES.md [hunk 1]
```diff
@@ -24,3 +24,6 @@ are ready, not on every PR.
 8. Verify it works.
 
 9. Allow the PR to be merged.
+
+
+[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/cluster/addons/dns/kube2sky/RELEASES.md?pixel)]()
```
Comment (thockin) [inline cluster/addons/dns/kube2sky/RELEASES.md]: Unrelated to this PR?

File: examples/limitrange/README.md [hunk 2]
```diff
@@ -1 +1,4 @@
 Please refer to this [doc](https://github.com/GoogleCloudPlatform/kubernetes/blob/620af168920b773ade28e27211ad684903a1db21/docs/design/admission_control_limit_range.md#kubectl).
+
+
+[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/examples/limitrange/README.md?pixel)]()
```
Comment (thockin) [inline examples/limitrange/README.md]: unrelated

File: pkg/proxy/port_allocator.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+func newRangeAllocator(r util.PortRange) PortAllocator {
+	if r.Base == 0 || r.Size == 0 {
+		panic("illegal argument: may not specify an empty port range")
+	}
+	sz := uint(r.Size)
+	ports := make(chan uint, sz)
+
+	// build a queue of shuffled ports
```
Comment (thockin) [inline pkg/proxy/port_allocator.go]: This really feels like doing it the hard way.  And the expensive way.  You've turned an N-bit bitmap into a map of N 64 bit ints and a channel buffering N 64 bit ints.

Why not just keep a bitmap and pick a random port on each allocation, retrying or scanning forward when you collide?

**Comment by thockin** [comment]
The idea is fine, thanks

File: cluster/addons/dns/kube2sky/RELEASES.md (hunk 1 above)
Comment (jdef) [inline cluster/addons/dns/kube2sky/RELEASES.md]: yes, unrelated. travis-ci was reporting issues that prevented it from building the code so I patched this up so that builds could complete. I can back this out if desired.

File: pkg/proxy/port_allocator.go (hunk 3 above)
Comment (jdef) [inline pkg/proxy/port_allocator.go]: Actually, this approach (map+chan) really simplifies allocation (don't need random+retry+scan) with a minor storage penalty. I can refactor to use a bitset (math/big) to save more space.

Comment (thockin) [inline pkg/proxy/port_allocator.go]: I guess it's not so terrible to keep a list of ints, but I would still
change the chan to buffer only 1 or 2 or 16 ints and feed it from a
goroutine, keeping state in a bitmap.  Randomly pick a bit in the bitmap,
if not free scan forward until you find a free one, claim it, write that to
the chan.

Comment (jdef) [inline pkg/proxy/port_allocator.go]: Thanks for the feedback. I've updated the code to use a random+sequential
search in a goroutine that fills a much smaller ports buf.

File: cmd/kube-proxy/app/server.go [hunk 4]
```diff
@@ -70,6 +71,7 @@ func (s *ProxyServer) AddFlags(fs *pflag.FlagSet) {
 	fs.IntVar(&s.OOMScoreAdj, "oom-score-adj", s.OOMScoreAdj, "The oom_score_adj value for kube-proxy process. Values must be within the range [-1000, 1000]")
 	fs.StringVar(&s.ResourceContainer, "resource-container", s.ResourceContainer, "Absolute name of the resource-only container to create and run the Kube-proxy in (Default: /kube-proxy).")
 	fs.StringVar(&s.Kubeconfig, "kubeconfig", s.Kubeconfig, "Path to kubeconfig file with authorization information (the master location is set by the master flag).")
+	fs.Var(&s.PortRange, "proxy-port-range", "Range of host ports that may be consumed in order to proxy service traffic.")
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: Commnt on the default behavior, please

File: pkg/proxy/port_allocator.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+// nextFreePort finds a free port, first attempting a random search. if a random search fails
+// then the port range is scanned sequentially until either a port is found or the scan completes
+// unsuccessfully. an unsuccessful scan returns a port of -1.
+func (r *rangeAllocator) nextFreePort() int {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// search randomly
+	for i := 0; i < randomAttempts; i++ {
```
Comment (thockin) [inline pkg/proxy/port_allocator.go]: I've had the same concern about other allocators - this feels more complex than needs to be.  Why not pick a random, if that random is busy, scan forward for the next opening.  If you loop all the way around, error.  Rather than randomly trying some arbitrary number of times.

File: pkg/proxy/port_allocator_test.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	r.Set("0-0")
+	defer func() {
+		if rv := recover(); rv == nil {
+			t.Fatalf("expected panic because of empty port range: %+v", r)
+		}
+	}()
+	_ = newRangeAllocator(*r)
+}
+
+func TestRangeAllocator_FullyAllocated(t *testing.T) {
```
Comment (thockin) [inline pkg/proxy/port_allocator_test.go]: Go doesn't use underscores.

File: pkg/proxy/port_allocator_test.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	// allocate the ports again
+	rports := make([]int, 100, 100)
+	for i := 0; i < 100; i++ {
+		rports[i], err = a.AllocateNext()
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}
+
+	if reflect.DeepEqual(ports, rports) {
```
Comment (thockin) [inline pkg/proxy/port_allocator_test.go]: This is a flake waiting to happen.  Can you instead release them in a deterministic order (e.g. backwards or some offset)?

File: pkg/proxy/proxier.go [hunk 8]
```diff
@@ -96,11 +97,17 @@ func NewProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptables.In
...
 	glog.V(2).Infof("Setting proxy IP to %v and initializing iptables", hostIP)
-	return createProxier(loadBalancer, listenIP, iptables, hostIP)
+	return createProxier(loadBalancer, listenIP, iptables, hostIP, proxyPorts)
 }
 
-func createProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptables.Interface, hostIP net.IP) (*Proxier, error) {
+func createProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptables.Interface, hostIP net.IP, proxyPorts PortAllocator) (*Proxier, error) {
+	// convenient to pass nil for tests..
+	if proxyPorts == nil {
+		proxyPorts = defaultPortAllocator
```
Comment (thockin) [inline pkg/proxy/proxier.go]: rather than referencing what should be a static variable (except go doesn't have those), why not call newAllocator() with a 0 range

**Comment by thockin** [comment]
LGTM but for a few nits.  We're code-freezing like, tomorrow, so it would be nice to get this locked down if you want it in for 1.0

Also, please edit the title and squash your commits.

**Comment by jdef** [comment]
Will address these concerns tomorrow morning, is that soon enough to make 1.0?

**Comment by thockin** [comment]
I think that PRs in this state of review can make it in, as long as it
doesn't sit for another week.

File: pkg/proxy/port_allocator_test.go (hunk 7 above)
Comment (jdef) [inline pkg/proxy/port_allocator_test.go]: done

**Comment by jdef** [comment]
Comments have been addressed PTAL

**Comment by jdef** [comment]
PTAL @thockin

**Comment by erictune** [comment]
@thockin what is the status of this PR re 1.0?

**Comment by thockin** [comment]
This was in flight before freeze, I just need to final review I think.  @jdef - needs rebase

File: pkg/proxy/port_allocator.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+// Release is a noop
+func (r *randomAllocator) Release(_ int) {
+	// noop
+}
+
+// newAllocator builds PortAllocator for a given PortRange. If the PortRange is empty
+// then a random port allocator is returned; otherwise, a new range-based allocator
+// is returned.
+func newAllocator(r util.PortRange) PortAllocator {
```
Comment (thockin) [inline pkg/proxy/port_allocator.go]: This needs a name more like "newPortAllocator" - newAllocator is not obvious enough for the entire scope of the proxy

File: pkg/proxy/proxier.go [hunk 10]
```diff
@@ -184,6 +192,10 @@ func (proxier *Proxier) stopProxy(service ServicePortName, info *serviceInfo) er
 
 // This assumes proxier.mu is locked.
 func (proxier *Proxier) stopProxyInternal(service ServicePortName, info *serviceInfo) error {
+	defer func() {
```
Comment (thockin) [inline pkg/proxy/proxier.go]: Why do this by defer rather than just linear code?

Comment (jdef) [inline pkg/proxy/proxier.go]: At the time I was thinking along the lines of a "finally" clause. Probably
doesn't matter here though. I'll make it linear.

**Comment by jdef** [comment]
rebase and impl of code review comments complete.

**Comment by thockin** [comment]
LGTM, thanks

**Comment by thockin** [comment]
What's the CLA situation?

**Comment by jdef** [comment]
I'm working on this for mesosphere and they've signed the CLA

**Comment by thockin** [comment]
I don't find Mesosphere in the corporate CLA list...

**Comment by jdef** [comment]
I double-checked and Tobi (@guenter) signed it recently on behalf of
Mesosphere.

**Comment by thockin** [comment]
Hmm, I don't have that in the DB that I can search.  Do you have a
confirmation or something I can reference?  How recent was this?

**Comment by jdef** [comment]
I've reached out to some folks internally about this. In the meantime I've
re-signed the commit with my mesosphere.io email.

**Comment by thockin** [comment]
Thanks,

I hate to be a stickler, but we really need it to be in order, even if that is a manual verification.

**Comment by jdef** [comment]
I heard back from @guenter, he said that he just signed it again.
